File: a2ui_agui_mcpapps_mcptoolkit/gemini-enterprise-a2a/a2ui_payloads.py
"""Deterministic A2UI v0.8 payloads for Gemini Enterprise."""

from __future__ import annotations

from typing import Any
from uuid import uuid4


def review_messages(
    recommendations: list[dict[str, Any]],
    approval_id: str,
) -> list[dict[str, Any]]:
    surface_id = f"inventory-transfer-review-{uuid4()}"
    children = ["title", "source"]
    components: list[dict[str, Any]] = [
        _component(
            "root",
            "Column",
            {"children": {"explicitList": children}},
        ),
        _text(
            "title",
            f"{len(recommendations)} governed inventory transfer recommendation(s)",
            "h2",
        ),
        _text(
            "source",
            "Live results from the Oracle Database MCP Java Toolkit",
            "caption",
        ),
    ]

    for index, recommendation in enumerate(recommendations, start=1):
        card_id = f"recommendation-card-{index}"
        content_id = f"recommendation-content-{index}"
        approve_id = f"approve-{index}"
        approve_text_id = f"approve-text-{index}"
        children.append(card_id)
        components.extend(
            [
                _component(card_id, "Card", {"child": content_id}),
                _component(
                    content_id,
                    "Column",
                    {
                        "children": {
                            "explicitList": [
                                f"name-{index}",
                                f"route-{index}",
                                f"risk-{index}",
                                f"rationale-{index}",
                                approve_id,
                            ]
                        }
                    },
                ),
                _text(
                    f"name-{index}",
                    f"{recommendation['sku']} - {recommendation['productName']}",
                    "h3",
                ),
                _text(
                    f"route-{index}",
                    (
                        f"{recommendation['sourceLocationCode']} to "
                        f"{recommendation['targetLocationCode']}: "
                        f"{recommendation['recommendedTransferQuantity']} units"
                    ),
                ),
                _text(
                    f"risk-{index}",
                    (
                        f"{recommendation['riskLevel']} stockout risk "
                        f"{recommendation['stockoutRiskScore']}"
                    ),
                ),
                _text(f"rationale-{index}", recommendation["rationale"]),
                _component(
                    approve_id,
                    "Button",
                    {
                        "child": approve_text_id,
                        "primary": True,
                        "action": {
                            "name": "approveInventoryTransfer",
                            "context": [
                                {
                                    "key": "approvalId",
                                    "value": {"literalString": approval_id},
                                },
                                {
                                    "key": "recommendationId",
                                    "value": {
                                        "literalString": recommendation[
                                            "recommendationId"
                                        ]
                                    },
                                },
                                {
                                    "key": "approvalNotes",
                                    "value": {"path": "/approvalNotes"},
                                },
                            ],
                        },
                    },
                ),
                _text(approve_text_id, "Approve this exact transfer"),
            ]
        )

    children.extend(["approval-notes", "cancel-review", "cancel-text"])
    components.extend(
        [
            _component(
                "approval-notes",
                "TextField",
                {
                    "label": {"literalString": "Approval notes"},
                    "text": {"path": "/approvalNotes"},
                    "textFieldType": "longText",
                },
            ),
            _component(
                "cancel-review",
                "Button",
                {
                    "child": "cancel-text",
                    "primary": False,
                    "action": {
                        "name": "rejectInventoryTransferReview",
                        "context": [
                            {
                                "key": "approvalId",
                                "value": {"literalString": approval_id},
                            }
                        ],
                    },
                },
            ),
            _text("cancel-text", "Cancel review without writing"),
        ]
    )
    return [
        {
            "beginRendering": {
                "surfaceId": surface_id,
                "root": "root",
            }
        },
        {
            "surfaceUpdate": {
                "surfaceId": surface_id,
                "components": components,
            }
        },
        {
            "dataModelUpdate": {
                "surfaceId": surface_id,
                "contents": [
                    {
                        "key": "approvalNotes",
                        "valueString": (
                            "Approve the database-recommended transfer "
                            "to reduce stockout exposure."
                        ),
                    }
                ],
            }
        },
    ]


def result_messages(title: str, detail: str) -> list[dict[str, Any]]:
    surface_id = f"inventory-transfer-result-{uuid4()}"
    return [
        {
            "beginRendering": {
                "surfaceId": surface_id,
                "root": "root",
            }
        },
        {
            "surfaceUpdate": {
                "surfaceId": surface_id,
                "components": [
                    _component(
                        "root",
                        "Card",
                        {"child": "content"},
                    ),
                    _component(
                        "content",
                        "Column",
                        {"children": {"explicitList": ["title", "detail"]}},
                    ),
                    _text("title", title, "h2"),
                    _text("detail", detail),
                ],
            }
        },
    ]


def _component(
    component_id: str,
    component_type: str,
    properties: dict[str, Any],
) -> dict[str, Any]:
    return {
        "id": component_id,
        "component": {component_type: properties},
    }


def _text(
    component_id: str,
    value: str,
    usage_hint: str | None = None,
) -> dict[str, Any]:
    properties: dict[str, Any] = {
        "text": {"literalString": str(value)}
    }
    if usage_hint:
        properties["usageHint"] = usage_hint
    return _component(component_id, "Text", properties)

File: a2ui_mcpapps_mcptoolkit/agent-service/src/main/java/com/oracle/demo/interactiveai/A2uiPayloads.java
package com.oracle.demo.interactiveai;

import java.util.List;
import java.util.Map;

public final class A2uiPayloads {
    public static final String VERSION = "v0.9.1";
    public static final String CATALOG =
            "https://a2ui.org/specification/v0_9_1/catalogs/basic/catalog.json";

    private A2uiPayloads() {
    }

    public static Map<String, Object> createSurface() {
        return Map.of("version", VERSION, "createSurface", Map.of(
                "surfaceId", "inventory-transfer-review",
                "catalogId", CATALOG,
                "sendDataModel", true,
                "theme", Map.of("primaryColor", "#c74634")));
    }

    public static Map<String, Object> components() {
        List<Map<String, Object>> components = List.of(
                Map.of(
                        "id", "root",
                        "component", "Column",
                        "children", List.of(
                                "banner",
                                "recommendationList",
                                "notes",
                                "buttons")),
                Map.of(
                        "id", "banner",
                        "component", "Text",
                        "text", Map.of("path", "/summary")),
                Map.of(
                        "id", "recommendationList",
                        "component", "List",
                        "children", Map.of(
                                "path", "/recommendations",
                                "componentId", "recommendationCard")),
                Map.of(
                        "id", "recommendationCard",
                        "component", "Card",
                        "child", "recommendationText"),
                Map.of(
                        "id", "recommendationText",
                        "component", "Text",
                        "text", Map.of(
                                "call", "formatString",
                                "args", Map.of(
                                        "value",
                                        "${sku}: move ${recommendedTransferQuantity} units from ${sourceLocationCode} to ${targetLocationCode}"))),
                Map.of(
                        "id", "notes",
                        "component", "TextField",
                        "label", "Transfer approval notes",
                        "value", Map.of("path", "/form/approvalNotes")),
                Map.of(
                        "id", "buttons",
                        "component", "Row",
                        "children", List.of("confirm", "cancel")),
                Map.of(
                        "id", "confirm",
                        "component", "Button",
                        "text", "Approve inventory transfer",
                        "variant", "primary",
                        "action", Map.of("event", Map.of(
                                "name", "approve_inventory_transfer"))),
                Map.of(
                        "id", "cancel",
                        "component", "Button",
                        "text", "Cancel",
                        "variant", "borderless",
                        "action", Map.of("event", Map.of(
                                "name", "reject_inventory_transfer"))));
        return Map.of(
                "version", VERSION,
                "updateComponents", Map.of(
                        "surfaceId", "inventory-transfer-review",
                        "components", components));
    }

    public static Map<String, Object> data(
            List<TransferRecommendation> recommendations,
            String approvalId,
            boolean writesAllowed) {
        List<Map<String, Object>> rows =
                recommendations.stream().map(Json::recommendationMap).toList();
        return Map.of(
                "version", VERSION,
                "updateDataModel", Map.of(
                        "surfaceId", "inventory-transfer-review",
                        "path", "/",
                        "value", Map.of(
                                "summary",
                                recommendations.size()
                                        + " governed inventory transfer recommendation(s) require review.",
                                "recommendations", rows,
                                "approvalId", approvalId,
                                "writesAllowed", writesAllowed,
                                "form", Map.of(
                                        "approvalNotes",
                                        "Approve the database-recommended transfer to reduce stockout exposure."))));
    }
}
